File: python/tests/environmental/bme680_test_linux.py
import os
from periph.transport.i2c_linux import I2CTransport
from periph.chips.environmental.bme680 import BME680Minimal, BME680Full

passed = 0
failed = 0

I2C_BUS = int(os.environ.get('LINUX_I2C_BUS', '1'))
I2C_ADDR = int(os.environ.get('I2C_ADDR', '0x76'), 16)

transport = I2CTransport(I2C_BUS, I2C_ADDR)

bme = BME680Minimal(transport)

t = bme.temperature()
if t >= -40 and t <= 85:
    print('PASS temperature_range')
    passed += 1
else:
    print('FAIL temperature_range: got {}'.format(t))
    failed += 1

p = bme.pressure()
if p >= 300 and p <= 1100:
    print('PASS pressure_range')
    passed += 1
else:
    print('FAIL pressure_range: got {}'.format(p))
    failed += 1

h = bme.humidity()
if h >= 0 and h <= 100:
    print('PASS humidity_range')
    passed += 1
else:
    print('FAIL humidity_range: got {}'.format(h))
    failed += 1

bme_full = BME680Full(transport)
if bme_full._osrs_t == 1 and bme_full._osrs_p == 1 and bme_full._osrs_h == 1:
    print('PASS default_oversampling')
    passed += 1
else:
    print('FAIL default_oversampling')
    failed += 1

bme_full.set_oversampling(BME680Full.OSRS_X4, BME680Full.OSRS_X2, BME680Full.OSRS_X1)
if bme_full._osrs_t == 3 and bme_full._osrs_p == 2 and bme_full._osrs_h == 1:
    print('PASS set_oversampling')
    passed += 1
else:
    print('FAIL set_oversampling')
    failed += 1

cid = bme_full.chip_id()
if cid == 0x61:
    print('PASS chip_id')
    passed += 1
else:
    print('FAIL chip_id: expected 0x61, got 0x{:02X}'.format(cid))
    failed += 1

t2, p2, h2, g2 = bme_full.read_all()
if t2 >= -40 and t2 <= 85 and p2 >= 300 and p2 <= 1100 and h2 >= 0 and h2 <= 100:
    print('PASS read_all')
    passed += 1
else:
    print('FAIL read_all: T={} P={} H={}'.format(t2, p2, h2))
    failed += 1

bme_full.reset()
print('PASS reset')
passed += 1

transport.close()
print('===DONE: {} passed, {} failed==='.format(passed, failed))

File: cpp/tests/environmental/bme680_test_linux/bme680_test_linux.cpp
#ifndef TEST_I2C_BUS
#define TEST_I2C_BUS 1
#endif
#ifndef TEST_ADDR
#define TEST_ADDR 0x76
#endif

#include <stdio.h>
#include "I2CTransportLinux.h"
#include "BME680.h"

static int passed = 0, failed = 0;

static void check_true(bool cond, const char *label) {
    if (cond) { printf("PASS %s\n", label); passed++; }
    else       { printf("FAIL %s\n", label); failed++; }
}

int main() {
    I2CTransportLinux transport(TEST_I2C_BUS, TEST_ADDR);

    BME680Minimal bme(transport);

    float t = bme.temperature();
    check_true(t >= -40.0f && t <= 85.0f, "temperature_range");

    float p = bme.pressure();
    check_true(p >= 300.0f && p <= 1100.0f, "pressure_range");

    float h = bme.humidity();
    check_true(h >= 0.0f && h <= 100.0f, "humidity_range");

    BME680Full bme_full(transport);

    bme_full.set_oversampling(BME680Full::OSRS_X4, BME680Full::OSRS_X2, BME680Full::OSRS_X1);
    check_true(bme_full._osrs_t == 3 && bme_full._osrs_p == 2 && bme_full._osrs_h == 1, "set_oversampling");

    check_true(bme_full.chip_id() == 0x61, "chip_id");

    float t2, p2, h2, g2;
    bme_full.read_all(t2, p2, h2, g2);
    check_true(t2 >= -40.0f && t2 <= 85.0f && p2 >= 300.0f && p2 <= 1100.0f && h2 >= 0.0f && h2 <= 100.0f, "read_all");

    bme_full.reset();
    check_true(true, "reset");

    printf("===DONE: %d passed, %d failed===\n", passed, failed);
    return failed == 0 ? 0 : 1;
}
